public/js/controllers/address_controller.js

```diff
@@ -80,36 +83,189 @@ function plotGraph (processedData, otherOptions) {
+      $(this).addClass('matching-hash')
+    } else {
+      $(this).removeClass('matching-hash')
+    }
+  })
+}
+
+class TurboQuery {
+  constructor (turbolinks) {
+    var ta = this
+    ta.replaceTimer = 0
+    ta.appendTimer = 0
+    ta.turbolinks = turbolinks || Turbolinks || false
+    if (!ta.turbolinks) {
+      console.error('No passed or global Turbolinks instance detected. TurboQuery requires Turbolinks.')
+      return
+    }
+    // These are timer callbacks. Bind them to the TurboQuery instance.
+    ta.replaceHistory = ta._replaceHistory.bind(ta)
+    ta.appendHistory = ta._appendHistory.bind(ta)
+    ta.url = Url(window.location.href, true)
+  }
+
+  replaceHref () {
+    // Rerouting through timer to prevent spamming.
+    // Turbolinks blocks replacement if frequency too high.
+    if (this.replaceTimer === 0) {
+      this.replaceTimer = setTimeout(this.replaceHistory, 250)
+    }
+  }
+
+  toHref () {
+    if (this.appendTimer === 0) {
+      this.appendTimer = setTimeout(this.appendHistory, 250)
+    }
+  }
+
+  _replaceHistory () {
+    // see https://github.com/turbolinks/turbolinks/issues/219. This also works:
+    // window.history.replaceState(window.history.state, this.addr, this.url.href)
+    this.turbolinks.controller.replaceHistoryWithLocationAndRestorationIdentifier(this.turbolinks.Location.wrap(this.url.href), this.turbolinks.uuid())
+    this.replaceTimer = 0
+  }
+
+  _appendHistory () {
+    // same as replaceHref, but creates a new entry in history for navigating
+    // with the browsers forward and back buttons. May still not work because of
+    // TurboLinks caching behavior, I think.
+    this.turbolinks.controller.pushHistoryWithLocationAndRestorationIdentifier(this.turbolinks.Location.wrap(this.url.href), this.turbolinks.uuid())
+    this.appendTimer = 0
+  }
+
+  replace (query) {
+    this.url.set('query', this.filteredQuery(query))
+    this.replaceHref()
+  }
+
+  to (query) {
+    this.url.set('query', this.filteredQuery(query))
+    this.toHref()
+  }
+
+  filteredQuery (query) {
+    var filtered = {}
+    Object.keys(query).forEach(function (key) {
+      var v = query[key]
+      if (typeof v === 'undefined' || v === null) return
+      filtered[key] = v
+    })
+    return filtered
+  }
+
+  update (target) {
+    // Update projects the current query parameters onto the given template.
+    return this.constructor.project(target, this.parsed)
+  }
+
+  get parsed () {
+    return this.url.query
+  }
+
+  get count () {
+    return Object.keys(this.url.query).length
+  }
+
+  get (key) {
+    if (this.url.query.hasOwnProperty(key)) {
+      return TurboQuery.parseValue(this.url.query[key])
+    }
+    return null
+  }
+
+  static parseValue (v) {
+    switch (v) {
+      case 'null':
+        return null
+      case '':
+        return null
+      case 'undefined':
+        return null
+      case 'false':
+        return false
+      case 'true':
+        return true
```

Comment:
What happens with `?q=1`? Seems like that should be a valid way to say true as well.

Reply:
That will parse as the integer `1`, which it "truthy". Similarly `?q=0` would parse to the integer `0`, which is "falsy".